### PR TITLE
feat/813 - notify speech-bargein-detected and dtmf-bargein-detected events

### PR DESCRIPTION
**Pull request by rammohan-kore**
when the prompt message is in progress, send speech-bargein-detected or dtmf-bargein-detected verb:status

### CODE REVIEW

**Comment by davehorton** [comment]
@rammohan-kore this needs to be rebased to latest main

**Comment by rammohan-kore** [comment]
> @rammohan-kore this needs to be rebased to latest main

merged main branch into this feature branch

File: lib/tasks/gather.js [hunk 1]
```diff
@@ -324,6 +324,7 @@ class TaskGather extends SttTask {
     if (this.dtmfBargein) {
       this._killAudio(cs);
       this.emit('dtmf', evt);
+      this.notifyStatus({event: 'dtmf-bargein-detected', ...evt});
```
[the remark this answers is not in the document]
Comment (rammohan-kore) [reply lib/tasks/gather.js:327]: When a prompt is being played, if user presses any keys on the dialpad on his phone, after the audio is  killed the dtmf-bargein-detected is sent once

[the remark this answers is not in the document]
Comment (rammohan-kore) [reply lib/tasks/gather.js:327]: Here is what I tested, when a prompt is being played and in between user presses a key, then the audio is getting killed and the dtmf-bargein-detected event is sent, however, if user presses other keys (after audio is killed), it is sending the dtmf-bargein-detected. I should be checking this.playComplete before sending ?

**Comment by davehorton** [comment]
I would test this example:
- gather with nested say, input dtmf and numDigits: 4
- user enters 1, 2, 3, 4
- confirm dtmf-bargein-detected is sent once, only when dtmf "1" is detected

**Comment by rammohan-kore** [comment]
> I would test this example:
> 
> * gather with nested say, input dtmf and numDigits: 4
> * user enters 1, 2, 3, 4
> * confirm dtmf-bargein-detected is sent once, only when dtmf "1" is detected

Yes @davehorton the event is sent multiple times, if I add if (!this.playComplete) it is sent only once

**Comment by rammohan-kore** [comment]
Should we change  the if condition in _killAudio  from  if (this.hasFillerNoise || (!this.sayTask && !this.playTask && this.bargein))  to     if (this.hasFillerNoise || (!this.sayTask && !this.playTask && (this.bargein | this.dtmfBargin))) 

without this change the configuration, this.playComplete is not set and multiple digits are sent even after audio is killed

**Comment by davehorton** [comment]
No, it seems to me that `if` statement is only triggered when there is neither a nested say nor play.  I would think instead in _killAudio we should add this final line at the end of the function
```
    if (this.playTask && !this.playTask.killed) {
      this.playTask.removeAllListeners('playDone');
      this.playTask.kill(cs);
      this.playTask = null;
    }
    this.playComplete = true;
```

**Comment by rammohan-kore** [comment]
> No, it seems to me that `if` statement is only triggered when there is neither a nested say nor play. I would think instead in _killAudio we should add this final line at the end of the function
> 
> ```
>     if (this.playTask && !this.playTask.killed) {
>       this.playTask.removeAllListeners('playDone');
>       this.playTask.kill(cs);
>       this.playTask = null;
>     }
>     this.playComplete = true;
> ```

after adding this.playComplete = true at the end of _killAudio multiple digits are not sending dtmf-bargein-detected event

File: lib/tasks/gather.js [hunk 2]
```diff
@@ -322,6 +322,9 @@ class TaskGather extends SttTask {
     clearTimeout(this.interDigitTimer);
     let resolved = false;
     if (this.dtmfBargein) {
+      if (!this.playComplete) {
```
Comment (davehorton) [inline lib/tasks/gather.js:325]: please change to a single condition
```
if (this.dtmfBargein && !this.playComplete) this.notifyStatus..
```

**Review by davehorton — CHANGES_REQUESTED** [review]

File: lib/tasks/gather.js (hunk 2 above)
Comment (rammohan-kore) [reply lib/tasks/gather.js]: added !this.playComplete to the main if condition

**Comment by vdharashive** [comment]
@rammohan-kore can you complete this please

**Comment by davehorton** [comment]
please rebase to main

**Comment by rammohan-kore** [comment]
> please rebase to main

merged main into this feat/813 branch

File: lib/tasks/gather.js [hunk 3]
```diff
@@ -321,7 +321,8 @@ class TaskGather extends SttTask {
     this.logger.debug(evt, 'TaskGather:_onDtmf');
     clearTimeout(this.interDigitTimer);
     let resolved = false;
-    if (this.dtmfBargein) {
+    if (this.dtmfBargein && !this.playComplete) {
```
Comment (davehorton) [inline lib/tasks/gather.js:324]: this doesnt look correct to me.  Given this code, once the play is complete we stop emitting dtmf events.  I understand you want to send a notify when the first dtmf is detected but you have included other side affects.

**Review by davehorton — CHANGES_REQUESTED** [review]

File: lib/tasks/gather.js (hunk 3 above)
Comment (rammohan-kore) [reply lib/tasks/gather.js]: if condition corrected and also referring the 0.2.9 version of time-series

**Review by davehorton — CHANGES_REQUESTED** [review]
In this latest commit, it appears as if you manually modified package-lock.json ?  I see a change to time-series@0.2.9 in the dependencies section, but then down below I see references to time-series@.0.28.

What you need to do is update package.json, as you have done, and then run `npm install` to generate the package-lock.json.  You should never be editing the package-lock.json file directly

**Comment by rammohan-kore** [comment]
> In this latest commit, it appears as if you manually modified package-lock.json ? I see a change to time-series@0.2.9 in the dependencies section, but then down below I see references to time-series@.0.28.
> 
> What you need to do is update package.json, as you have done, and then run `npm install` to generate the package-lock.json. You should never be editing the package-lock.json file directly

My bad, will ensure going forward, I have generated the package-lock.json using npm -i

**Review by davehorton — APPROVED** [review]

**Pull request merged**